map/src/main/kotlin/edu/artic/map/MapViewModel.kt

```diff
@@ -0,0 +1,200 @@
+
+    val floor: Subject<Int> = BehaviorSubject.createDefault(1)
+    val zoomLevel: Subject<MapZoomLevel> = BehaviorSubject.create()
+
+    val currentFloor: Int
+        get() {
+            val castFloor = (floor as BehaviorSubject)
+            return if (castFloor.hasValue()) {
+                castFloor.value
+            } else {
+                1
+            }
+        }
+
+    val currentZoomLevel: MapZoomLevel
+        get() {
+            val castLevel = (zoomLevel as BehaviorSubject)
+            return if (castLevel.hasValue()) {
+                castLevel.value
+            } else {
+                MapZoomLevel.One
+            }
+        }
+
+    init {
+        val allAmenities = mapAnnotationDao
+                .getAmenitiesOnMap()
+                .map { annotationList ->
+                    val list = mutableListOf<MapItem.Annotation>()
+                    annotationList.forEach { annotation ->
+                        val floor = annotation.floor
+                                .let {
+                                    it?.toInt() ?: 1
+                                }
+                        list.add(MapItem.Annotation(annotation, floor))
+                    }
+                    return@map list
+
+                }
+
+        val allBuildingNames = mapAnnotationDao
+                .getBuildingNamesOnMap()
+                .map { annotationList ->
+                    val list = mutableListOf<MapItem.Annotation>()
+                    annotationList.forEach { annotation ->
+                        val floor = annotation.floor
+                                .let {
+                                    it?.toInt() ?: 1
+                                }
+                        list.add(MapItem.Annotation(annotation, floor))
+                    }
+                    return@map list
+
+                }
+
+        Observable.combineLatest(
+                allAmenities.asObservable(),
+                allBuildingNames.asObservable(),
+                BiFunction<List<MapItem.Annotation>, List<MapItem.Annotation>, List<MapItem.Annotation>>
+                { amenities, buildingNames ->
+                    val list = mutableListOf<MapItem.Annotation>()
+                    list.addAll(amenities)
+                    list.addAll(buildingNames)
+                    return@BiFunction list
+                }).bindTo(alwaysVisibleAnnotations)
+                .disposedBy(disposeBag)
+
+
+        setupZoomLevelOneBinds()
+        setupZoomLevelTwoBinds()
+        setupZoomLevelThreeBinds()
+
+    }
+
+
+    fun setupZoomLevelOneBinds() {
+        Observable.combineLatest(
+                zoomLevel.distinctUntilChanged().filter { it === MapZoomLevel.One },
+                floor.distinctUntilChanged(),
+                alwaysVisibleAnnotations,
+                Function3<MapZoomLevel, Int, List<MapItem.Annotation>, List<MapItem<*>>>
```

Comment:
not sure we need explicit signatures here. are you not using `RXKotlin`? `Observables.combineLatest()`

Reply:
yes we need explicit signature. I've never been able to get it working without that signature

Reply:
I was using Obsevable instead of Observables
fixing

Reply:
done